README.md
<p align="center">
    <img src="https://raw.githubusercontent.com/raycharius/clean-redirect/main/resources/images/logo-horizontal.png" alt="Logo" width="600px">
</p>

<p align="center">
    <h3 align="center">All your redirect logic wrapped into one single redirect</h3>
</p>

<p align="center">
    Force HTTPS, remove trailing slash, to lower case, naked to www, and your own custom logic, as one redirect.
    <br />
    <br />
    <a href="#floppy_disk--installation">Get Started</a>
    ·
    <a href="https://github.com/raycharius/clean-redirect/issues">Request Feature</a>
    ·
    <a href="https://github.com/raycharius/clean-redirect/issues">Report Bug</a>
  </p>
</p>

***

[![npm](https://img.shields.io/npm/v/clean-redirect?color=bright-green)](https://www.npmjs.com/package/clean-redirect)
![NPM](https://img.shields.io/npm/l/clean-redirect?color=bright-green)
[![codecov](https://codecov.io/gh/raycharius/clean-redirect/branch/main/graph/badge.svg)](https://codecov.io/gh/raycharius/clean-redirect)
[![Maintainability](https://api.codeclimate.com/v1/badges/a9aecc5560c08e013ec8/maintainability)](https://codeclimate.com/github/raycharius/clean-redirect/maintainability)

When it comes to redirects, it's easy to end up with a redirect chain. First your server forces HTTPS. Redirects to the naked domain. Then removes the trailing slash. And only then does the request hit your backend, where it is met with redirects dependent on your business logic. **Clean Redirect** is a middleware for Express that helps you avoid redirect chains by keeping all of the redirect rules in one place, redirecting only once they've all been applied. 

While it may be desirable to keep some of those redirects (such as force HTTPS) on the web server, there are cases when they need to happen on the backend, such as when hosting an application on Heroku or similar infrastructure.

## :zap: &nbsp; Features

* Enforce HTTPS.
* Redirect from naked domain to WWW and vice versa.
* Remove trailing slash from path.
* Convert path to lowercase.
* The ability to pass in a function with your own business logic.
* A lot of configurable options to keep it as simple or governed as you need it.

## :floppy_disk: &nbsp; Installation 

#### Using NPM: 

``` bash
npm install --save clean-redirect
```

#### Using Yarn: 

``` bash
yarn add clean-redirect
```
## :mag: &nbsp; Example

```javascript

const express = require('express');
const cleanRedirect = require('clean-redirect');

const app = express();

const getCustomRedirectLocation = (req, res, redirector) => {
  if (redirector.path === '/store') {
    redirector
      .setPath('/marketplace')
      .setPersistQueryString(false)
      .setRedirectCode(302);
  }
};

const redirector = cleanRedirect({
  forceHttps: true,
  toWww: true,
  pathToLowerCase: true,
  removeTrailingSlash: true,
  persistQueryString: true,
  customRedirects: getCustomRedirectLocation,
  callNextOnRedirect: true,
  redirectCode: 301,
});

app.use(redirector);

// http://yourapp.com/SUBSCRIBE/?email=johndoe@gmail.com => 301
// https://www.yourapp.com/subscribe?email=johndoe@gmail.com

// http://yourapp.com/STORE/?showBundles=true => 302
// https://www.yourapp.com/marketplace

```

## :space_invader: &nbsp; Usage

**Clean Redirect** exposes a single middleware function that accepts multiple options, some of which are used to configure the redirect logic, others which are used to govern the behavior of the middleware.

### Options 

| Option                | Type     | Required | Description                                                                                                                                                                                                       |
|-----------------------|----------|----------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `forceHttps`          | Boolean  | No       | Redirects non-secure HTTP requests to HTTPS. Defaults to `false`.                                                                                                                                                 |
| `toWww`               | Boolean  | No       | Redirects requests to the naked (root) domain to the WWW domain. Defaults to  `false`. Both `toWww` and `toNaked` cannot set to `true` at the same time.                                                          |
| `toNaked`             | Boolean  | No       | Redirects requests to the WWW domain to the naked (root) domain. Defaults to  `false`. Both `toWww` and `toNaked` cannot set to `true` at the same time.                                                          |
| `pathToLowerCase`     | Boolean  | No       | Redirects requests with capital letters in the path to the same path, but lowercase. Defaults to  `false`.                                                                                                        |
| `removeTrailingSlash` | Boolean  | No       | Redirects requests with a trailing slash to the same path, without a slash. Defaults to  `false`.                                                                                                                 |
| `persistQueryString`  | Boolean  | No       | When set, the query string from the original request will be persisted to the redirect location. Defaults to  `false` (default behavior for Express).                                                             |
| `redirectCode`        | Integer  | No       | Sets the redirect code to be used by default. Valid values are `301` and `302`. Defaults to `302`.                                                                                                                |
| `callNextOnRedirect`  | Boolean  | No       | When set to `true`, the `next()` function will be called in the middleware after the redirect has taken place. Defaults to `false`.                                                                               |
| `deferRedirectToNext` | Boolean  | No       | When set to `true`, `res.redirect()` is not called in the middleware. Instead, an instance of CleanRedirect is saved to `res.locals` and available for further governance in the middleware that follows.         |
| `customRedirects`     | Function | No       | This is a function that accepts three arguments (`req`, `res`, `redirector`) and is called before calling `res.redirect()`. Pass this in to include custom redirect logic using conditions and setters and getters for the `CleanRedirect` object (see below). |

Ideally, this middleware should come before any other redirect-related logic, globally, using the `app.use()` method. However, you can also pass different configurations into different areas of your application.

### Methods for `CleanRedirect`

There are two cases in which there is direct access to the object (an instance of `CleanRedirect`) that stores all the original and target URL data:

* Within the function passed to the `customRedirects` option.
* In later middleware, through the `res.locals.cleanRedirect` property, when passing a value of `true` to `deferRedirectToNext`.

In these situations, there are getters and setters available to further modify the configuration of the redirect and values of the target URL which will be passed into the `res.redirect()` method.

#### Getter Methods

| Name                  | Type     | Description                                                                                                                                                                                                       |
|-----------------------|----------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `protocol`            | String   | Get the target protocol.
| `hostname`            | String   | Get the target hostname.
| `path`                | String   | Get the target path.
| `queryString`         | String   | Get the target queryString.
| `hash`                | String   | Get the target hash.
| `requiresRedirect`    | Boolean  | Compares the original URL data and the mutated values to decide if a redirect is needed or not. 
| `redirectUrl`         | String   | Returns the full string that will be passed into the `res.redirect()` method.
| `redirectCode`        | Integer  | Get the configured redirect code (301 or 302).

#### Setter Methods

All setter methods return `this` and are thus chainable.

| Method                             | Description                                                                                                                                                                                                       |
|------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `setProtocol(string)`              | Set the target protocol.
| `setHostname(string)`              | Set the target hostname.
| `setPath(string)`                  | Set the target path.
| `setQueryString(string)`           | Set the target queryString.
| `setHash(string)`                  | Set the target hash.
| `setPersistQueryString(boolean)`   | Set the option to persist the query string on the target URL. 
| `setRedirectCode(integer)`         | Set the redirect code to be passed into the `res.redirect()` method.
| `setPathOverride(string)`          | Set an explicit value for the path passed into the `res.redirect()` method. Can be utilized to use the path relative and back redirects [supported by Express](https://expressjs.com/en/api.html#res.redirect). When set, all target URL data will be ignored, and the value passed as an override will be directly passed into `res.redirect()`.

## :fire: &nbsp; Acknowledgements

<img src="https://cdn.dribbble.com/users/683635/avatars/normal/ee2c7c826bfe244b573d145376fe0b5a.png?1510328842" alt="@ft502" width="24" height="24" valign="bottom" /> Alexey Chernyshov ([@ft502](https://dribbble.com/ft502) on Dribbble) - For the Clean Redirect logo!

## :black_nib: &nbsp; Author

<img src="https://github.com/raycharius.png" alt="@raycharius" width="24" height="24" valign="bottom" /> Ray East ([@raycharius](https://github.com/raycharius)) - Technical Product Manager by Day, NodeJS Developer by Night

CHANGELOG.md
# Change Log

## v1.0.0 - 2020-11-20

* Stable release. 
* Update documentation in README.md
* Add JSDoc documentation to source code.

## v0.1.1 - 2020-11-12

* Fix bug with protocol being taken always from `req.protocol`, instead of taking into account the `x-forwarded-protocol` header.

## v0.1.0 - 2020-11-11

* Initial Beta Release


test.js
const CleanRedirect = require('./src/clean-redirect');

const requestData = {
  method: 'GET',
  protocol: 'http',
  hostname: 'macpaw.com',
  uri: '/cleanmyMAC/?hello=world&goodbye=blue-sky#fuckery',
};

const config = {
  forceHttps: true,
  toWww: true,
  toNaked: false,
  toLowerCase: true,
  removeTrailingSlash: true,
  persistQueryString: true,
  persistHash: true,
  redirectType: 301,
  customRedirects: (cleanRedirect) => cleanRedirect.setPersistHash(false),
};

const cleanRedirect = new CleanRedirect(requestData, config);

console.log(cleanRedirect);
console.log(cleanRedirect.requiresRedirect);
console.log(cleanRedirect.getRedirectUrl());


tests/validator.spec.js
const validator = require('../src/validator');

describe('Validators – Check that function properly', () => {
  test('validateCleanRedirectConfig', () => {
    expect(() => validator.validateCleanRedirectConfig({ toWww: true, toNaked: true, redirectCode: 301 })).toThrow();
    expect(() => validator.validateCleanRedirectConfig({ toWww: true, toNaked: false, redirectCode: 303 })).toThrow();

    validator.validateCleanRedirectConfig({ toWww: true, toNaked: false, redirectCode: 301 });
  });

  test('validateRedirectCode', () => {
    expect(() => validator.validateRedirectCode(303)).toThrow();
  });

  test('validateCleanRedirectUrlArgs', () => {
    expect(() => validator.validateCleanRedirectUrlArgs({ protocol: undefined, hostname, uri })).toThrow();
    expect(() => validator.validateCleanRedirectUrlArgs({ protocol, hostname: undefined, uri })).toThrow();
    expect(() => validator.validateCleanRedirectUrlArgs({ protocol, hostname, uri: undefined })).toThrow();

    validator.validateCleanRedirectUrlArgs({ protocol: 'http', hostname: 'gobiden.com', uri: '/make/democracy/great/again' });
  });

  test('validateBooleanCleanRedirectConfigOptions', () => {
    expect(() => validator.validateBooleanCleanRedirectConfigOptions({ toWww: false, forceHttps: 'no' })).toThrow();
    expect(() => validator.validateBooleanCleanRedirectConfigOptions({ toWww: false, forceHttps: 'no' })).toThrow();

    validator.validateBooleanCleanRedirectConfigOptions({ toWww: false, forceHttps: true });
  });

  test('validateIsString', () => {
    expect(() => validator.validateIsString(301)).toThrow();

    validator.validateIsString('hello');
  });

  test('validateIsBoolean', () => {
    expect(() => validator.validateIsBoolean('hello')).toThrow();

    validator.validateIsBoolean(true);
  });

  test('validateCustomRedirects', () => {
    expect(() => validator.validateCustomRedirects('hello')).toThrow();

    validator.validateCustomRedirects(() => true);
  });
});

tests/clean-redirect.spec.js
const CleanRedirect = require('../src/clean-redirect');
const CleanRedirectUrl = require('../src/clean-redirect-url');
const {
  protocol,
  hostname,
  path,
  queryString,
  hash,
  uri,
  url,
  hostnameNoWww,
  uriTrailingSlash,
  uriUpperCase,
  uriUpperCaseTrailingSlash,
} = require('./mocks/clean-redirect.mock');

describe('CleanRedirect – Instantiation works correctly', () => {
  test('Instantiation is successful, and sourceUrl and targetUrl are instances of CleanRedirectUrl, and sourceUrl values are valid', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri });

    expect(cleanRedirect.sourceUrl).toBeInstanceOf(CleanRedirectUrl);
    expect(cleanRedirect.targetUrl).toBeInstanceOf(CleanRedirectUrl);
    expect(cleanRedirect.sourceUrl).toEqual(cleanRedirect.targetUrl);
    expect(cleanRedirect.sourceUrl.protocol).toEqual(protocol);
    expect(cleanRedirect.sourceUrl.hostname).toEqual(hostname);
    expect(cleanRedirect.sourceUrl.path).toEqual(path);
    expect(cleanRedirect.sourceUrl.queryString).toEqual(queryString);
    expect(cleanRedirect.sourceUrl.uri).toEqual(uri);
    expect(cleanRedirect.sourceUrl.url).toEqual(url);
  });

  test('Instantiation results in pathOverride having value of null', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri });

    expect(cleanRedirect.pathOverride).toEqual(null);
  });

  test('Instantiation with no options sets all options to false, with redirect type of 302', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri });

    expect(cleanRedirect.config.forceHttps).toEqual(false);
    expect(cleanRedirect.config.toWww).toEqual(false);
    expect(cleanRedirect.config.toNaked).toEqual(false);
    expect(cleanRedirect.config.pathToLowerCase).toEqual(false);
    expect(cleanRedirect.config.removeTrailingSlash).toEqual(false);
    expect(cleanRedirect.config.persistQueryString).toEqual(false);
    expect(cleanRedirect.config.alwaysPassFullUrl).toEqual(false);
    expect(cleanRedirect.config.redirectCode).toEqual(302);
  });

  test('Instantiation with all options set to false sets options to false, with redirect type of 302', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri }, {
      forceHttps: false,
      toWww: false,
      toNaked: false,
      pathToLowerCase: false,
      removeTrailingSlash: false,
      persistQueryString: false,
      alwaysPassFullUrl: false,
    });

    expect(cleanRedirect.config.forceHttps).toEqual(false);
    expect(cleanRedirect.config.toWww).toEqual(false);
    expect(cleanRedirect.config.toNaked).toEqual(false);
    expect(cleanRedirect.config.pathToLowerCase).toEqual(false);
    expect(cleanRedirect.config.removeTrailingSlash).toEqual(false);
    expect(cleanRedirect.config.persistQueryString).toEqual(false);
    expect(cleanRedirect.config.alwaysPassFullUrl).toEqual(false);
    expect(cleanRedirect.config.redirectCode).toEqual(302);
  });

  test('Instantiation with all options set to true sets options to true (except toWww), with redirect type of 302', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri }, {
      forceHttps: true,
      toWww: false,
      toNaked: true,
      pathToLowerCase: true,
      removeTrailingSlash: true,
      persistQueryString: true,
      alwaysPassFullUrl: true,
    });

    expect(cleanRedirect.config.forceHttps).toEqual(true);
    expect(cleanRedirect.config.toWww).toEqual(false);
    expect(cleanRedirect.config.toNaked).toEqual(true);
    expect(cleanRedirect.config.pathToLowerCase).toEqual(true);
    expect(cleanRedirect.config.removeTrailingSlash).toEqual(true);
    expect(cleanRedirect.config.persistQueryString).toEqual(true);
    expect(cleanRedirect.config.alwaysPassFullUrl).toEqual(true);
    expect(cleanRedirect.config.redirectCode).toEqual(302);
  });

  test('Instantiation with toWww set to true results in a value of true', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri }, { toWww: true });

    expect(cleanRedirect.config.toWww).toEqual(true);
  });

  test('Instantiation with both toWww and toNaked as true throw an error', () => {
    expect(() => new CleanRedirect({ protocol, hostname, uri }, { toNaked: true, toWww: true })).toThrow();
  });

  test('Instantiation with non-boolean value to config throws an error', () => {
    expect(() => new CleanRedirect({ protocol, hostname, uri }, { toNaked: 'invalid' })).toThrow();
  });

  test('Instantiation with redirectCode set to 302 results in a 302 redirectCode', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri }, { redirectCode: 302 });

    expect(cleanRedirect.config.redirectCode).toEqual(302);
  });

  test('Instantiation with redirectCode set to 301 results in a 301 redirectCode', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri }, { redirectCode: 301 });

    expect(cleanRedirect.config.redirectCode).toEqual(301);
  });

  test('Instantiation with redirectCode set to something other than 301 or 302 results in an error being thrown', () => {
    expect(() => new CleanRedirect({ protocol, hostname, uri }, { redirectCode: 'invalid' })).toThrow();
  });
});

describe('CleanRedirect – Parsers work correctly based on passed options', () => {
  test('With all options set to false, no values are mutated', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri });

    expect(cleanRedirect.targetUrl.protocol).toEqual(protocol);
    expect(cleanRedirect.targetUrl.hostname).toEqual(hostname);
    expect(cleanRedirect.targetUrl.path).toEqual(path);
    expect(cleanRedirect.targetUrl.queryString).toEqual(queryString);
    expect(cleanRedirect.targetUrl.uri).toEqual(uri);
    expect(cleanRedirect.targetUrl.url).toEqual(url);
  });

  test('With forceHttps set to true, a protocol value of http is mutated into https, redirect is necessary, etc.', () => {
    const cleanRedirect = new CleanRedirect({ protocol: 'http', hostname, uri }, { forceHttps: true });
    const expectedUrl = `https://${hostname}${path}${queryString}`;
    const expectedRedirectUrl = `https://${hostname}${path}`;

    expect(cleanRedirect.targetUrl.protocol).toEqual('https');
    expect(cleanRedirect.targetUrl.hostname).toEqual(hostname);
    expect(cleanRedirect.targetUrl.path).toEqual(path);
    expect(cleanRedirect.targetUrl.queryString).toEqual(queryString);
    expect(cleanRedirect.targetUrl.uri).toEqual(uri);
    expect(cleanRedirect.targetUrl.url).toEqual(expectedUrl);
    expect(cleanRedirect.requiresRedirect).toEqual(true);
    expect(cleanRedirect.redirectUrl).toEqual(expectedRedirectUrl);
  });

  test('With toWww set to true, a hostname value starting without www. is mutated into hostname with www, redirect is necessary, etc.', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname: hostnameNoWww, uri }, { toWww: true });
    const expectedUrl = `${protocol}://${hostname}${path}${queryString}`;
    const expectedRedirectUrl = `${hostname}${path}`;

    expect(cleanRedirect.targetUrl.protocol).toEqual(protocol);
    expect(cleanRedirect.targetUrl.hostname).toEqual(hostname);
    expect(cleanRedirect.targetUrl.path).toEqual(path);
    expect(cleanRedirect.targetUrl.queryString).toEqual(queryString);
    expect(cleanRedirect.targetUrl.uri).toEqual(uri);
    expect(cleanRedirect.targetUrl.url).toEqual(expectedUrl);
    expect(cleanRedirect.requiresRedirect).toEqual(true);
    expect(cleanRedirect.redirectUrl).toEqual(expectedRedirectUrl);
  });

  test('With toNaked set to true, a hostname value starting with www. is mutated into hostname without www, redirect is necessary, etc.', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri }, { toNaked: true });
    const expectedUrl = `${protocol}://${hostnameNoWww}${path}${queryString}`;
    const expectedRedirectUrl = `${hostnameNoWww}${path}`;

    expect(cleanRedirect.targetUrl.protocol).toEqual(protocol);
    expect(cleanRedirect.targetUrl.hostname).toEqual(hostnameNoWww);
    expect(cleanRedirect.targetUrl.path).toEqual(path);
    expect(cleanRedirect.targetUrl.queryString).toEqual(queryString);
    expect(cleanRedirect.targetUrl.uri).toEqual(uri);
    expect(cleanRedirect.targetUrl.url).toEqual(expectedUrl);
    expect(cleanRedirect.requiresRedirect).toEqual(true);
    expect(cleanRedirect.redirectUrl).toEqual(expectedRedirectUrl);
  });

  test('With removeTrailingSlash set to true, a path with a trailing slash is mutated into path without one, redirect is necessary, etc.', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri: uriTrailingSlash }, { removeTrailingSlash: true });
    const expectedUrl = `${protocol}://${hostname}${path}${queryString}`;
    const expectedRedirectUrl = path;

    expect(cleanRedirect.targetUrl.protocol).toEqual(protocol);
    expect(cleanRedirect.targetUrl.hostname).toEqual(hostname);
    expect(cleanRedirect.targetUrl.path).toEqual(path);
    expect(cleanRedirect.targetUrl.queryString).toEqual(queryString);
    expect(cleanRedirect.targetUrl.uri).toEqual(uri);
    expect(cleanRedirect.targetUrl.url).toEqual(expectedUrl);
    expect(cleanRedirect.requiresRedirect).toEqual(true);
    expect(cleanRedirect.redirectUrl).toEqual(expectedRedirectUrl);
  });

  test('With pathToLowerCase set to true, a path with a upper case letters is mutated into path without them, redirect is necessary, etc.', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri: uriUpperCase }, { pathToLowerCase: true });
    const expectedUrl = `${protocol}://${hostname}${path}${queryString}`;
    const expectedRedirectUrl = path;

    expect(cleanRedirect.targetUrl.protocol).toEqual(protocol);
    expect(cleanRedirect.targetUrl.hostname).toEqual(hostname);
    expect(cleanRedirect.targetUrl.path).toEqual(path);
    expect(cleanRedirect.targetUrl.queryString).toEqual(queryString);
    expect(cleanRedirect.targetUrl.uri).toEqual(uri);
    expect(cleanRedirect.targetUrl.url).toEqual(expectedUrl);
    expect(cleanRedirect.requiresRedirect).toEqual(true);
    expect(cleanRedirect.redirectUrl).toEqual(expectedRedirectUrl);
  });

  test('With pathToLowerCase and removeTrailingSlash set to true, the path is lower case, without trailing slash, redirect is necessary, etc.', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri: uriUpperCaseTrailingSlash }, { removeTrailingSlash: true, pathToLowerCase: true });
    const expectedUrl = `${protocol}://${hostname}${path}${queryString}`;
    const expectedRedirectUrl = path;

    expect(cleanRedirect.targetUrl.protocol).toEqual(protocol);
    expect(cleanRedirect.targetUrl.hostname).toEqual(hostname);
    expect(cleanRedirect.targetUrl.path).toEqual(path);
    expect(cleanRedirect.targetUrl.queryString).toEqual(queryString);
    expect(cleanRedirect.targetUrl.uri).toEqual(uri);
    expect(cleanRedirect.targetUrl.url).toEqual(expectedUrl);
    expect(cleanRedirect.requiresRedirect).toEqual(true);
    expect(cleanRedirect.redirectUrl).toEqual(expectedRedirectUrl);
  });

  test('With persistQueryString set to true and toLowerCase, the path is lower case, query string is available in redirect URL, redirect is necessary, etc.', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri: uriUpperCase }, { persistQueryString: true, pathToLowerCase: true });
    const expectedUrl = `${protocol}://${hostname}${path}${queryString}`;
    const expectedRedirectUrl = `${path}${queryString}`;

    expect(cleanRedirect.targetUrl.protocol).toEqual(protocol);
    expect(cleanRedirect.targetUrl.hostname).toEqual(hostname);
    expect(cleanRedirect.targetUrl.path).toEqual(path);
    expect(cleanRedirect.targetUrl.queryString).toEqual(queryString);
    expect(cleanRedirect.targetUrl.uri).toEqual(uri);
    expect(cleanRedirect.targetUrl.url).toEqual(expectedUrl);
    expect(cleanRedirect.requiresRedirect).toEqual(true);
    expect(cleanRedirect.redirectUrl).toEqual(expectedRedirectUrl);
  });

  test('With alwaysPassFullUrl set to true and toLowerCase, the path is lower case, URL is full, without query string, redirect is necessary, etc.', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri: uriUpperCase }, { alwaysPassFullUrl: true, pathToLowerCase: true });
    const expectedUrl = `${protocol}://${hostname}${path}${queryString}`;
    const expectedRedirectUrl = `${protocol}://${hostname}${path}`;

    expect(cleanRedirect.targetUrl.protocol).toEqual(protocol);
    expect(cleanRedirect.targetUrl.hostname).toEqual(hostname);
    expect(cleanRedirect.targetUrl.path).toEqual(path);
    expect(cleanRedirect.targetUrl.queryString).toEqual(queryString);
    expect(cleanRedirect.targetUrl.uri).toEqual(uri);
    expect(cleanRedirect.targetUrl.url).toEqual(expectedUrl);
    expect(cleanRedirect.requiresRedirect).toEqual(true);
    expect(cleanRedirect.redirectUrl).toEqual(expectedRedirectUrl);
  });

  test('With alwaysPassFullUrl, toLowerCase, persistQueryString, the path is lower case, URL is full, with query string, redirect is necessary, etc.', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri: uriUpperCase }, { alwaysPassFullUrl: true, pathToLowerCase: true, persistQueryString: true });
    const expectedUrl = `${protocol}://${hostname}${path}${queryString}`;
    const expectedRedirectUrl = `${protocol}://${hostname}${path}${queryString}`;

    expect(cleanRedirect.targetUrl.protocol).toEqual(protocol);
    expect(cleanRedirect.targetUrl.hostname).toEqual(hostname);
    expect(cleanRedirect.targetUrl.path).toEqual(path);
    expect(cleanRedirect.targetUrl.queryString).toEqual(queryString);
    expect(cleanRedirect.targetUrl.uri).toEqual(uri);
    expect(cleanRedirect.targetUrl.url).toEqual(expectedUrl);
    expect(cleanRedirect.requiresRedirect).toEqual(true);
    expect(cleanRedirect.redirectUrl).toEqual(expectedRedirectUrl);
  });

  test('All URL settings set to true should result in correct path', () => {
    const cleanRedirect = new CleanRedirect({ protocol: 'http', hostname, uri: uriUpperCaseTrailingSlash }, {
      forceHttps: true,
      toNaked: true,
      pathToLowerCase: true,
      removeTrailingSlash: true,
      persistQueryString: true,
    });
    const expectedUrl = `https://${hostnameNoWww}${path}${queryString}`;

    expect(cleanRedirect.targetUrl.protocol).toEqual('https');
    expect(cleanRedirect.targetUrl.hostname).toEqual(hostnameNoWww);
    expect(cleanRedirect.targetUrl.path).toEqual(path);
    expect(cleanRedirect.targetUrl.queryString).toEqual(queryString);
    expect(cleanRedirect.targetUrl.uri).toEqual(uri);
    expect(cleanRedirect.targetUrl.url).toEqual(expectedUrl);
    expect(cleanRedirect.requiresRedirect).toEqual(true);
    expect(cleanRedirect.redirectUrl).toEqual(expectedUrl);
  });
});

describe('CleanRedirect – Getter methods work correctly', () => {
  test('Getters retrieve the values of the targetUrl object', () => {
    const cleanRedirect = new CleanRedirect({ protocol: 'http', hostname, uri: uriUpperCaseTrailingSlash }, {
      forceHttps: true,
      toNaked: true,
      pathToLowerCase: true,
      removeTrailingSlash: true,
      persistQueryString: true,
    });

    cleanRedirect.setHash(hash);

    expect(cleanRedirect.protocol).toEqual(cleanRedirect.targetUrl.protocol);
    expect(cleanRedirect.hostname).toEqual(cleanRedirect.targetUrl.hostname);
    expect(cleanRedirect.path).toEqual(cleanRedirect.targetUrl.path);
    expect(cleanRedirect.queryString).toEqual(cleanRedirect.targetUrl.queryString);
    expect(cleanRedirect.hash).toEqual(cleanRedirect.targetUrl.hash);

  });

  test('Getter for redirect type returns redirect type in config object', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri }, { redirectCode: 302 });

    expect(cleanRedirect.redirectCode).toEqual(302);
    expect(cleanRedirect.redirectCode).toEqual(cleanRedirect.config.redirectCode);
  });
});

describe('CleanRedirect – Setter methods work correctly', () => {
  test('Setters for URL parts set the values of the targetUrl object', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri }, { persistQueryString: true });

    cleanRedirect.setProtocol('voting');

    expect(cleanRedirect.targetUrl.protocol).toEqual('voting');

    cleanRedirect.setHostname('usa-elections.com');

    expect(cleanRedirect.targetUrl.hostname).toEqual('usa-elections.com');

    cleanRedirect.setPath('/the/path/to/redemption');

    expect(cleanRedirect.targetUrl.path).toEqual('/the/path/to/redemption');

    cleanRedirect.setQueryString('?year=2020');

    expect(cleanRedirect.targetUrl.queryString).toEqual('?year=2020');

    cleanRedirect.setHash('#goamerica');

    expect(cleanRedirect.targetUrl.hash).toEqual('#goamerica');

    expect(cleanRedirect.requiresRedirect).toEqual(true);
    expect(cleanRedirect.redirectUrl).toEqual('voting://usa-elections.com/the/path/to/redemption?year=2020#goamerica');
  });

  test('Setters for config object set the values correctly', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri });

    cleanRedirect.setPersistQueryString(true);

    expect(cleanRedirect.config.persistQueryString).toEqual(true);

    cleanRedirect.setRedirectCode(301);

    expect(cleanRedirect.config.redirectCode).toEqual(301);

    cleanRedirect.setPathOverride('..');

    expect(cleanRedirect.pathOverride).toEqual('..');
  });

  test('Once set, pathOverride is returned by the getRedirectUrl method', () => {
    const cleanRedirect = new CleanRedirect({ protocol, hostname, uri });

    cleanRedirect.setPathOverride('..');

    expect(cleanRedirect.pathOverride).toEqual('..');
    expect(cleanRedirect.requiresRedirect).toEqual(true);
    expect(cleanRedirect.redirectUrl).toEqual('..');
  });
});


src/validator.js
const CleanRedirectError = require('./clean-redirect-error');

const validateBooleanCleanRedirectConfigOptions = ({ redirectCode, ...config }) => {
  Object.keys(config).forEach((key) => {
    const value = config[key];

    if (typeof value !== 'boolean') {
      throw new CleanRedirectError(`Expected a value of type 'boolean' for options '${key}'. Instead received type '${typeof typeof value}'`);
    }
  });
};

const validateRedirectCode = (code) => {
  const validRedirectCodes = [301, 302];

  if (!validRedirectCodes.includes(code)) {
    throw new CleanRedirectError('Option \'redirectCode\' must have an integer value of either 301 or 302');
  }
};

const validateCleanRedirectConfig = (config) => {
  validateBooleanCleanRedirectConfigOptions(config);

  if (config.toWww && config.toNaked) {
    throw new CleanRedirectError('Both \'toWww\' and \'toNaked\' cannot be simultaneously set to true.');
  }

  validateRedirectCode(config.redirectCode);
};

const validateCleanRedirectUrlArgs = (urlData) => {
  const keys = ['protocol', 'hostname', 'uri'];

  keys.forEach((key) => {
    const value = urlData[key];

    if (value === 'undefined' || typeof value !== 'string') {
      throw new CleanRedirectError(`CleanRedirectError requires a string value for parameter '${key}', received a value with type '${typeof value}'`);
    }
  });
};

const validateIsString = (string) => {
  if (typeof string !== 'string') {
    throw new CleanRedirectError(`Expected a value of type 'string'. Instead received type '${typeof string}'`);
  }
};

const validateIsBoolean = (bool) => {
  if (typeof bool !== 'boolean') {
    throw new CleanRedirectError(`Expected a value of type 'boolean'. Instead received type '${typeof bool}'`);
  }
};

const validateCustomRedirects = (customRedirects) => {
  if (customRedirects && typeof customRedirects !== 'function') {
    throw new CleanRedirectError('When set, the option \'customRedirects\' must either a function');
  }
};

module.exports = {
  validateCleanRedirectConfig,
  validateIsString,
  validateRedirectCode,
  validateIsBoolean,
  validateCustomRedirects,
  validateCleanRedirectUrlArgs,
  validateBooleanCleanRedirectConfigOptions,
};
